File: backend/src/apps/notification/models/notification.py
"""Notification ORM model."""
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING, Any, Optional

from sqlalchemy import JSON
from sqlmodel import Field, Relationship, SQLModel

if TYPE_CHECKING:
    from src.apps.iam.models.user import User


class NotificationType(str, Enum):
    INFO = "info"
    SUCCESS = "success"
    WARNING = "warning"
    ERROR = "error"
    PAYMENT = "payment"
    AUTH = "auth"
    SYSTEM = "system"


class Notification(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: int = Field(foreign_key="user.id", index=True)
    title: str = Field(max_length=255)
    body: str = Field(max_length=2000)
    type: NotificationType = Field(default=NotificationType.INFO)
    is_read: bool = Field(default=False)
    extra_data: Optional[Any] = Field(default=None, sa_type=JSON)
    created_at: datetime = Field(default_factory=datetime.now)

    # Relationships
    user: Optional["User"] = Relationship(back_populates="notifications")

File: mobile/lib/features/notifications/data/models/notification.dart
enum NotificationType {
  info,
  success,
  warning,
  error,
  payment,
  auth,
  system;

  static NotificationType fromString(String v) {
    return NotificationType.values.firstWhere(
      (e) => e.name == v,
      orElse: () => NotificationType.info,
    );
  }
}

class AppNotification {
  final String id;
  final String userId;
  final String title;
  final String body;
  final NotificationType type;
  final bool isRead;
  final dynamic extraData;
  final String createdAt;

  const AppNotification({
    required this.id,
    required this.userId,
    required this.title,
    required this.body,
    required this.type,
    required this.isRead,
    this.extraData,
    required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id'].toString(),
      userId: json['user_id']?.toString() ?? '',
      title: json['title'] as String? ?? '',
      body: json['body'] as String? ?? '',
      type: NotificationType.fromString(json['type'] as String? ?? 'info'),
      isRead: json['is_read'] as bool? ?? false,
      extraData: json['extra_data'],
      createdAt: json['created_at'] as String? ?? '',
    );
  }
}
